lib/virt_autotest/utils.pm

```diff
@@ -319,4 +320,48 @@ sub collect_virt_system_logs {
     upload_asset '/tmp/dumpxml.tar.gz';
 }
 
+# is_guest_online($guest) check if the given guests is online by probing for an open ssh port
+sub is_guest_online {
+    my $guest = shift;
+    return script_run("nmap $guest -PN -p ssh | grep open") == 0;
+}
+
+# Shutdown all guests and wait until they are shutdown
+sub shutdown_guests {
+    ## Reboot the guest to ensure the settings are applied
+    # Shutdown and start the guest because some might have the on_reboot=destroy policy still applied
+    script_run("virsh shutdown $_") foreach (keys %virt_autotest::common::guests);
+    # Wait until guests are terminated
+    wait_guests_shutdown();
+}
+
+# wait_guests_shutdown([$timeout]) waits for all guests to be shutdown
+sub wait_guests_shutdown {
+    my $retries = shift // 240;
+    # Note: Domain-0 is for xen only, but it does not hurt to exclude this also in kvm runs.
+    script_retry("! virsh list | grep -v Domain-0 | grep running", delay => 1, retry => $retries);
+}
+
+# wait_guest_online($guest, [$timeout]) waits until the given guests is online by probing for an open ssh port
+sub wait_guest_online {
+    my $guest   = shift;
+    my $retries = shift // 300;
+    # Wait until guest is reachable via ssh
+    script_retry("nmap $guest -PN -p ssh | grep open", delay => 1, retry => $retries);
+}
+
+# wait_guest_online([$timeout]) waits until all guests are online by calling wait_guest_online on them
+sub wait_all_guests_online {
```

Comment:
Is this `_guests` wrapper necessary? We anyways do `foreach` in test modules.

Reply:
Necessary? No, but I find a

    wait_all_guests_online();

more beautiful than

    wait_guest_online($_, $timeout) foreach (keys %virt_autotest::common::guests);

You think we should remove this one because it's redundant?

Reply:
Let me put it this way, sometimes we have also this:
```perl
foreach (keys %virt_autotest::common::guests) {
    script_run "ssh root\@$guest reboot";
    wait_guest_online($_);
}
```
I just find those _all function as too much fragmentation.